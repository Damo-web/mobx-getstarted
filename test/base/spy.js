"use strict"
var mobx = require("../../src/mobx.ts")
const utils = require("../utils/test-utils")

test("spy output", () => {
    var events = []

    var stop = mobx.spy(c => events.push(c))

    doStuff()

    stop()

    doStuff()

    events.forEach(ev => {
        delete ev.object
        delete ev.fn
        delete ev.time
    })

    expect(events).toMatchSnapshot()
})

function doStuff() {
    var a = mobx.observable.box(2)
    a.set(3)

    var b = mobx.observable({
        c: 4
    })
    b.c = 5
    mobx.extendObservable(b, { d: 6 })
    b.d = 7

    var e = mobx.observable([1, 2])
    e.push(3, 4)
    e.shift()
    e[2] = 5

    var f = mobx.observable.map({ g: 1 })
    f.delete("h")
    f.delete("g")
    f.set("i", 5)
    f.set("i", 6)

    var j = mobx.computed(() => a.get() * 2)

    var stop = mobx.autorun(() => {
        j.get()
    })

    a.set(4)

    mobx.transaction(function myTransaction() {
        a.set(5)
        a.set(6)
    })

    mobx
        .action("myTestAction", newValue => {
            a.set(newValue)
        })
        .call({}, 7)
}

test("spy error", () => {
    utils.supressConsole(() => {
        mobx._getGlobalState().mobxGuid = 0

        const a = mobx.observable({
            x: 2,
            get y() {
                if (this.x === 3) throw "Oops"
                return this.x * 2
            }
        })

        var events = []
        var stop = mobx.spy(c => events.push(c))

        var d = mobx.autorun(() => a.y, { name: "autorun" })

        a.x = 3

        events.forEach(x => {
            delete x.fn
            delete x.object
            delete x.time
        })

        expect(events).toMatchSnapshot()

        d()
        stop()
    })
})

test("spy stop listen from handler, #1459", () => {
    const stop = mobx.spy(() => stop())
    mobx.spy(() => {})
    doStuff()
})
