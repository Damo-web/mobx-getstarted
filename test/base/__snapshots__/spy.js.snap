// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spy error 1`] = `
Array [
  Object {
    "name": "autorun",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "name": "ObservableObject@1.y",
    "type": "compute",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "x",
    "name": "ObservableObject@1",
    "newValue": 3,
    "oldValue": 2,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "name": "ObservableObject@1.y",
    "type": "compute",
  },
  Object {
    "name": "autorun",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "error": "Oops",
    "message": "[mobx] Encountered an uncaught exception that was thrown by a reaction or observer component, in: 'Reaction[autorun]",
    "name": "autorun",
    "type": "error",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "spyReportEnd": true,
  },
]
`;

exports[`spy output 1`] = `
Array [
  Object {
    "name": "ObservableValue@1",
    "newValue": "2",
    "type": "create",
  },
  Object {
    "name": "ObservableValue@1",
    "newValue": 3,
    "oldValue": 2,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "c",
    "name": "ObservableObject@2",
    "newValue": 4,
    "spyReportStart": true,
    "type": "add",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "c",
    "name": "ObservableObject@2",
    "newValue": 5,
    "oldValue": 4,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "d",
    "name": "ObservableObject@2",
    "newValue": 6,
    "spyReportStart": true,
    "type": "add",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "d",
    "name": "ObservableObject@2",
    "newValue": 7,
    "oldValue": 6,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "added": Array [
      1,
      2,
    ],
    "addedCount": 2,
    "index": 0,
    "name": "ObservableArray@3",
    "removed": Array [],
    "removedCount": 0,
    "spyReportStart": true,
    "type": "splice",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "added": Array [
      3,
      4,
    ],
    "addedCount": 2,
    "index": 2,
    "name": "ObservableArray@3",
    "removed": Array [],
    "removedCount": 0,
    "spyReportStart": true,
    "type": "splice",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "added": Array [],
    "addedCount": 0,
    "index": 0,
    "name": "ObservableArray@3",
    "removed": Array [
      1,
    ],
    "removedCount": 1,
    "spyReportStart": true,
    "type": "splice",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "index": 2,
    "name": "ObservableArray@3",
    "newValue": 5,
    "oldValue": 4,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "g",
    "name": "ObservableMap@4",
    "newValue": 1,
    "spyReportStart": true,
    "type": "add",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "g",
    "name": "ObservableMap@4",
    "oldValue": 1,
    "spyReportStart": true,
    "type": "delete",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "i",
    "name": "ObservableMap@4",
    "newValue": 5,
    "spyReportStart": true,
    "type": "add",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "key": "i",
    "name": "ObservableMap@4",
    "newValue": 6,
    "oldValue": 5,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "Autorun@7",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "name": "ComputedValue@6",
    "type": "compute",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "ObservableValue@1",
    "newValue": 4,
    "oldValue": 3,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "name": "ComputedValue@6",
    "type": "compute",
  },
  Object {
    "name": "Autorun@7",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "ObservableValue@1",
    "newValue": 5,
    "oldValue": 4,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "ObservableValue@1",
    "newValue": 6,
    "oldValue": 5,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "ComputedValue@6",
    "type": "compute",
  },
  Object {
    "name": "Autorun@7",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "arguments": Array [
      7,
    ],
    "name": "myTestAction",
    "spyReportStart": true,
    "type": "action",
  },
  Object {
    "name": "ObservableValue@1",
    "newValue": 7,
    "oldValue": 6,
    "spyReportStart": true,
    "type": "update",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "name": "ComputedValue@6",
    "type": "compute",
  },
  Object {
    "name": "Autorun@7",
    "spyReportStart": true,
    "type": "reaction",
  },
  Object {
    "spyReportEnd": true,
  },
  Object {
    "spyReportEnd": true,
  },
]
`;
